### DISCUSSION TITLE
Problem when import Fanuc CRX25 robot into database

### DISCUSSION THREAD

**Discussion by WiesenDX** [visose/Robots#241] [Q&A] [answered]
I am trying to import the Fanuc-CRX25 robot into the plugin, and I have chosen an ABB robot to replace the values of the a and d parameters (using UR will result in an error).
However, due to the fact that the 4, 5, and 6 axes of the ABB robot are in a straight line, and there is two rotation among the 4, 5, and 6 axes of the Fanuc CEX25, after I replaced the new geometry and parameter values of a and d, it still seems like following the default TCP position of the original ABB robot (the 3 axis and target are in a straight line, and the true TCP cannot overlap with the target).
Link below has my replaced geometry, XML, and GH files for testing.

[https://wetransfer.com/downloads/ae19b85d7b6586d8e8e30486b8dd72a420230810153124/a5a463474b549d2f3d02c0be9d0677cf20230810153138/430c5a](https://wetransfer.com/downloads/ae19b85d7b6586d8e8e30486b8dd72a420230810153124/a5a463474b549d2f3d02c0be9d0677cf20230810153138/430c5a?trk=TRN_TDL_01&utm_campaign=TRN_TDL_01&utm_medium=email&utm_source=sendgrid)

Thank you very much!
![image](https://github.com/visose/Robots/assets/127881362/50d9836f-4b4c-4854-a273-73a6dc242d28)

**Comment by visose** [comment]
Fanuc is not supported by this plugin at the moment. If you're interested in sponsoring the work you can send an email to support@visose.com for a quote.

**Reply by visose** [reply]
What is likely happening in your case is that the kinematics solver used for ABB robots requires a spherical wrist, which doesn't seem like is the case.

**Reply by WiesenDX** [reply]
Yeah I think so, is there any robot in database whose wrist is like this?

**Reply by visose** [reply]
Not exactly the same, but you can try with UR. Franka uses has a solver that works with any kinematic chain but is hardcoded for Franka robots. You might have to change some code in the plugin to get it working.

**Reply by WiesenDX** [reply]
I think so, in this case I also need to change some code in the plugin not only change in XML file. But it's not possible for me to change code in the plugin right?

**Reply by visose** [reply]
Sure, you can fork this repo and have a go.

**Comment by visose** [comment, marked as answer]
Closing this as superseded by current Fanuc support. Fanuc is now a supported manufacturer in Robots, so do not spoof a Fanuc robot as ABB/UR in current versions. Start from an existing Fanuc library or create a Fanuc `.xml`/`.3dm` pair. If a specific CRX model still has nonstandard kinematics that the Fanuc solver cannot represent, that would need a focused new discussion or code changes.
